code/game/turfs/floors/floor_icon.dm

```diff
@@ -158,41 +196,41 @@
 	return global.flooring_cache[cache_key]
 
 /decl/flooring/proc/test_link(var/turf/origin, var/turf/opponent)
-	var/is_linked = FALSE
-	if(istype(origin) && istype(opponent))
-		//is_wall is true for wall turfs and for floors containing a low wall
-		if(opponent.is_wall())
-			if(wall_smooth == SMOOTH_ALL)
-				is_linked = TRUE
-		//If is_hole is true, then it's space or openspace
-		else if(opponent.is_open())
-			if(space_smooth == SMOOTH_ALL)
-				is_linked = TRUE
-
-		//If we get here then its a normal floor
-		else if (istype(opponent, /turf/floor))
-			var/turf/floor/floor_opponent = opponent
-			//If the floor is the same as us,then we're linked,
-			if (istype(src, floor_opponent.flooring))
-				is_linked = TRUE
-			else if (floor_smooth == SMOOTH_ALL)
-				is_linked = TRUE
-			else if (floor_smooth != SMOOTH_NONE)
-				//If we get here it must be using a whitelist or blacklist
-				if (floor_smooth == SMOOTH_WHITELIST)
-					if (flooring_whitelist[floor_opponent.flooring.type])
-						//Found a match on the typecache
-						is_linked = TRUE
-				else if(floor_smooth == SMOOTH_BLACKLIST)
-					is_linked = TRUE //Default to true for the blacklist, then make it false if a match comes up
-					if (flooring_blacklist[floor_opponent.flooring.type])
-						//Found a match on the typecache
-						is_linked = FALSE
-			//Check for window frames.
-			if (!is_linked && wall_smooth == SMOOTH_ALL)
-				if(locate(/obj/structure/wall_frame) in opponent)
-					is_linked = TRUE
-	return is_linked
+	if(!istype(origin) || !istype(opponent))
+		return FALSE
+
+	. = FALSE
+	//is_wall is true for wall turfs and for floors containing a low wall
+	if(opponent.is_wall())
+		if(wall_smooth == SMOOTH_ALL)
+			. = TRUE
+	//If is_hole is true, then it's space or openspace
+	else if(opponent.is_open())
+		if(space_smooth == SMOOTH_ALL)
+			. = TRUE
+
+	//If we get here then its a normal floor
+	else if (istype(opponent, /turf/floor))
+		var/turf/floor/floor_opponent = opponent
+		if (floor_smooth == SMOOTH_ALL)
+			. = TRUE
+		//If the floor is the same as us,then we're linked,
+		else if (floor_opponent.flooring?.type == neighbour_type)
```

Comment:
```suggestion
		else if (istype(floor_opponent.flooring, neighbour_type))
```
the point of this is so that path subtypes (like herringbone, running, etc) don't overlap each other. without this, cobble paths draw edges over brick paths, which is not great visually and a regression from current behavior